nodeJS/express/express_forms_validation.md

```diff
@@ -0,0 +1,565 @@
+- Using `express-validator` for validation and sanitization.
+- Best practices for form design and data integrity.
+
+### HTML forms overview
+
+First, a brief refresher on the actual HTML.
+
+Let's create a simple HTML form, with a single text field for collecting a full name, and its associated label element.
+
+The HTML structure would look something like this:
+
+```html
+<form action="/users/create" method="POST" >
+  <label for="fullName">Full Name:</label>
+  <input placeholder="John Doe" type="text" name="fullName" id="fullName">
+  <button type="submit">Submit</button>
+</form>
+```
+
+We have the HTML form element itself, with an action pointing to some resource on our server, and a method defined. Notice how the method corresponds to an HTTP verb, typically either `GET` or `POST`.
+
+The input needs a corresponding label element, which is what the end user actually sees on the page.
+
+Then, we have our input. It has an `id` and a `name` attribute, which must match the `for` attribute defined on our label above. The `name` attribute is especially important, as it defines how our input will be identified in the form data sent to our server.
+
+It has a type of `text` which tells the browser what type of widget to display, what values to accept, and is used for JavaScript and CSS selectors.
+
+The `type="submit"` button can be pressed by the user to upload the data entered into all other input elements to the server (in this case there's only one).
+
+The `form` attributes define how to communicate with the server:
+
+- `action`: The resource/URL where data is sent for processing when the form is submitted. If this is not set, or is an empty string, the form will be submitted back to the current page URL
+- `method`: Defines the HTTP method to use (`POST` or `GET`).
+
+`POST` is generally more secure because it keeps sensitive information out of the URL, which means they won't show up in server logs, and is the standard choice for creating or updating data on the server side. `GET` is for forms that don't modify data, such as search forms, or when you want the form submission to be bookmarkable or shareable via URL.
+
+#### Form handling process
+
+The form's `action` will point to some endpoint on our server, which will allow our controller the request. The controller then communicates with the database to handle the data.
+
+We then generate a new or updated view with the controller's response and redirect the client. This is known as the [Post/Redirect/Get (PRG)](https://en.wikipedia.org/wiki/Post/Redirect/Get) design pattern which helps prevent duplicate `POST` requests.
+
+### Validation and sanitization
+
+Before the data from a form is sent off to our server, we must first do two things
+
+- *Validation* ensures user input meets the specified criteria (e.g. required fields, correct format).
+- *Sanitization* cleans user input to prevent malicious data from being processed by removing or encoding potentially malicious characters.
```

Comment:
I'm a little iffy on the wording of this part. Currently, it reads like we all our form data must be *both* validated and sanitised. Of course we must validate all form data as necessary, but sanitisation (especially at receipt) is only needed if absolutely needed (as discussed before and changes made already).

e.g. outputting unescaped data is bad, but we handle that by escaping when outputting, not handling the sanitisation at receipt of input.

I can personally see this bit confusing learners who are told they *must* sanitise all form data, but then are shown examples where sanitisation does not occur and we only escape on output.